Atomic/sign.py

```diff
@@ -0,0 +1,89 @@
+            self.args.sign_key is None else self.args.sign_key
+
+        # Make sure the signature path exists
+        if not os.path.exists(signature_path):
+            raise ValueError("The path signature path {} "
+                             "does not exist".format(signature_path))
+
+        for sign_image in self.args.sign_targets:
+            local_inspect_info = self._inspect_image(image=sign_image)
+            sign_image_id = local_inspect_info['Id']
+            manifest = util.skopeo_inspect('docker://{}'.format(sign_image), args=['--raw'])
+
+            # Make sure the local image is the same as the image on the
+            # registry
+
+            if not ids_same(sign_image_id, manifest['config']['digest']):
+                raise ValueError("The local image {} differs from the "
+                                 "version on the registry".format(sign_image))
+            with tempfile.NamedTemporaryFile(mode="w", delete=False) as manifest_file:
+                manifest_file.write(str(manifest))
+                reg, image, tag = util.decompose(self.get_fq_name(local_inspect_info)) # pylint: disable=unused-variable
+
+
+                if self.args.output is None:
+                    sigstore_path = os.path.join(signature_path,
+                                                 "{}/{}@{}".format(reg,
+                                                                   image,
+                                                                   add_shaw(sign_image_id)))
+                    self.make_sig_dirs(sigstore_path)
+                    signature_int = self.get_sig_int(sigstore_path)
+                    fq_sig_path = os.path.join(sigstore_path, "signature-{}".format(signature_int))
+
+                else:
+                    fq_sig_path = os.path.join(self.args.output)
+                    self.make_sig_dirs(os.path.dirname(fq_sig_path))
+
+                util.skopeo_standalone_sign(sign_image, manifest_file.name,
+                                            self.get_fingerprint(signer), fq_sig_path)
+
+    def check_input_validity(self):
+        try:
+            for image in self.args.sign_targets:
+                self._is_image(image)
+        except AtomicError:
+            raise ValueError("{} is not a valid image".format(image))
+
+    @staticmethod
+    def get_fingerprint(signer):
+        # Need handling of errors still
+        cmd = ['gpg', '--no-permission-warning', '--with-colons', '--fingerprint', signer]
+        return_code, stdout, stderr = util.subp(cmd)
+        if return_code is not 0:
+            raise ValueError(stderr)
+        for line in stdout.splitlines():
+            if line.startswith('fpr:'):
+                return line.split(":")[9]
+
+    @staticmethod
+    def make_sig_dirs(sig_path):
+        if not os.path.exists(sig_path):
+            os.makedirs(sig_path)
+
+    @staticmethod
+    def get_sig_int(sig_path):
+        return len(os.listdir(sig_path)) + 1
```

Comment:
This feels brittle. An aborted operation might leave high-numbered signature files around, which are ignored by https://github.com/mtrmac/image/blob/docker-lookaside/docker/docker_image_dest.go (which stops at the lowest unused number). Or there might be any random files in there (OS X Finder metadata, whatever).

https://github.com/mtrmac/image/blob/docker-lookaside/docker/docker_image_dest.go enumerates 1, 2, … and stops at first nonexistent file. In there it is especially motivated by HTTP support, where there isn’t a universally easily supported way to list the directory anyway, AFAIK.